packages/openspp/src/Adaptor.js

```diff
@@ -918,6 +920,164 @@ export function removeFromGroup(group_id, individual_id) {
+ * @param {number} [offset=0] - offset searching
+ * @param {function} callback - An optional callback function
+ * @returns {Operation}
+ */
+export function searchServicePoint(domain, offset=0, callback=(s) => s) {
+  return async state => {
+    try {
+      let isDomain = true;
+      for (const element of domain) {
+        if (!Array.isArray(element)) {
+          isDomain = false;
+          break;
+        }
+      }
+      if (!isDomain) {
+        domain = [domain];
+      }
+      let servicePoints = await sppConnector.searchRead(
+        'spp.service.point',
+        domain,
+        [
+          'name',
+          'area_id',
+          'service_type_ids',
+          'program_id',
+          'phone_sanitized',
+          'is_contract_active',
+          'is_disabled',
+        ],
+        {
+          limit: 100,
+          offset: offset,
+          order: 'id desc'
+        },
+      );
+      for (let servicePoint of servicePoints) {
+        servicePoint.program_ids = servicePoint.program_id;
+        delete servicePoint.program_id;
+      }
+      if (servicePoints.length === 0) {
+        console.log(`⚠ Warning: Service point with domain=${domain} not found!`);
+        return state;
+      }
+      console.log(`ℹ Service point with domain=${domain} found!`);
+      const nextState = composeNextState(state, servicePoints);
+      return callback(nextState);
+    } catch (err) {
+      console.log(`✗ Error: ${err}`);
+      return state;
+    }
+  }
+}
+
+/**
+ * get area by id in OpenSPP
+ * @public
+ * @example
+ * getArea(1)
+ * @function
+ * @param {number} id - id of area
+ * @param {function} callback - An optional callback function
+ * @returns {Operation}
+ */
+export function getArea(id, callback=(s) => s) {
+  return async state => {
+    try {
+      const area = await sppConnector.read(
+        'spp.area',
+        id,
+        [
+          'parent_id',
+          'name',
+          'code',
+          'altnames',
+          'area_level',
+          'kind'
+        ],
+      );
+      if (area.length === 0) {
+        console.log(`⚠ Warning: Area ${id} not found!`);
+        return state;
+      }
+      console.log(`ℹ Area ${id} found!`);
+      const nextState = composeNextState(state, area);
+      return callback(nextState);
+    } catch (err) {
+      console.log(`✗ Error: ${err}`);
+      return state;
+    }
+  }
+}
+
+/**
+ * searching for service point in OpenSPP
+ * @public
+ * @example <caption>search without offset</caption>
+ * searchArea([["code", "=", "10732"]])
+ * @example <caption>search with offset</caption>
+ * searchArea([["kind", "=", 1]], 1)
+ * @function
+ * @param {Array} domain - searching domain
+ * @param {number} [offset=0] - offset searching
+ * @param {function} callback - An optional callback function
+ * @returns {Operation}
+ */
+export function searchArea(domain, offset=0, callback=(s) => s) {
+  return async state => {
+    try {
+      let isDomain = true;
```

Comment:
This block of code, which scans `domain` and maybe embeds it in an array, is repeated. It would be nice to factor it out into another function and just call it from here.
```
const resolveDomain = (domain) => {
       let isDomain = true;
      for (const element of domain) {
        if (!Array.isArray(element)) {
          isDomain = false;
          break;
        }
      }
      if (!isDomain) {
        domain = [domain];
      }
     return domain;
}

// then in your operation:
```
try {
   domain = resolveDomain(domain)
    const areas = await sppConnector.searchRead( ... )
}
```
I think this does the same thing but a bit more cleanly?
```
const resolveDomain = (domain) => {
    for (const element of domain) {
        if (!Array.isArray(element)) {
            return [domain]
        }
    }
}
```
We normally put functions like this into a `Utils.js` file but that's not strictly necessary. Also, functions like this are great to unit test!